// thirdweb/dist/cjs/react/web/ui/ConnectWallet/screens/Buy/WalletSelectorButton.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletSelectorButton = WalletSelectorButton;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_icons_1 = require("@radix-ui/react-icons");
const address_js_1 = require("../../../../../../utils/address.js");
const CustomThemeProvider_js_1 = require("../../../../../core/design-system/CustomThemeProvider.js");
const index_js_1 = require("../../../../../core/design-system/index.js");
const useConnectedWallets_js_1 = require("../../../../hooks/wallets/useConnectedWallets.js");
const WalletImage_js_1 = require("../../../components/WalletImage.js");
const basic_js_1 = require("../../../components/basic.js");
const buttons_js_1 = require("../../../components/buttons.js");
const text_js_1 = require("../../../components/text.js");
const GenericWalletIcon_js_1 = require("../../icons/GenericWalletIcon.js");
function WalletSelectorButton(props) {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    const connectedWallets = (0, useConnectedWallets_js_1.useConnectedWallets)();
    const walletId = props.walletId ||
        connectedWallets.find((x) => x.getAccount()?.address === props.address)?.id;
    return ((0, jsx_runtime_1.jsx)(basic_js_1.Container, { bg: "tertiaryBg", style: {
            borderRadius: index_js_1.radius.lg,
            border: `1px solid ${theme.colors.borderColor}`,
            ...props.containerStyle,
        }, children: (0, jsx_runtime_1.jsxs)(buttons_js_1.Button, { fullWidth: true, disabled: props.disabled, onClick: props.onClick, variant: "ghost", style: {
                justifyContent: "space-between",
                padding: index_js_1.spacing.sm,
                backgroundColor: theme.colors.tertiaryBg,
            }, gap: "sm", children: [(0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "row", center: "y", gap: "sm", color: "secondaryText", children: [walletId ? ((0, jsx_runtime_1.jsx)(WalletImage_js_1.WalletImage, { id: walletId, size: index_js_1.iconSize.md, client: props.client })) : ((0, jsx_runtime_1.jsx)(GenericWalletIcon_js_1.GenericWalletIcon, { size: index_js_1.iconSize.md })), (0, jsx_runtime_1.jsx)(text_js_1.Text, { size: "sm", color: "primaryText", children: (0, address_js_1.shortenAddress)(props.address) })] }), !props.disableChevron && ((0, jsx_runtime_1.jsx)(react_icons_1.ChevronDownIcon, { width: index_js_1.iconSize.sm, height: index_js_1.iconSize.sm, style: {
                        color: theme.colors.secondaryText,
                    } })), props.checked && ((0, jsx_runtime_1.jsx)(react_icons_1.CheckIcon, { width: index_js_1.iconSize.md, height: index_js_1.iconSize.md }))] }) }));
}
