// thirdweb/dist/esm/react/web/ui/ConnectWallet/screens/Buy/WalletSelectorButton.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { CheckIcon, ChevronDownIcon } from "@radix-ui/react-icons";
import { shortenAddress } from "../../../../../../utils/address.js";
import { useCustomTheme } from "../../../../../core/design-system/CustomThemeProvider.js";
import { iconSize, radius, spacing, } from "../../../../../core/design-system/index.js";
import { useConnectedWallets } from "../../../../hooks/wallets/useConnectedWallets.js";
import { WalletImage } from "../../../components/WalletImage.js";
import { Container } from "../../../components/basic.js";
import { Button } from "../../../components/buttons.js";
import { Text } from "../../../components/text.js";
import { GenericWalletIcon } from "../../icons/GenericWalletIcon.js";
export function WalletSelectorButton(props) {
    const theme = useCustomTheme();
    const connectedWallets = useConnectedWallets();
    const walletId = props.walletId ||
        connectedWallets.find((x) => x.getAccount()?.address === props.address)?.id;
    return (_jsx(Container, { bg: "tertiaryBg", style: {
            borderRadius: radius.lg,
            border: `1px solid ${theme.colors.borderColor}`,
            ...props.containerStyle,
        }, children: _jsxs(Button, { fullWidth: true, disabled: props.disabled, onClick: props.onClick, variant: "ghost", style: {
                justifyContent: "space-between",
                padding: spacing.sm,
                backgroundColor: theme.colors.tertiaryBg,
            }, gap: "sm", children: [_jsxs(Container, { flex: "row", center: "y", gap: "sm", color: "secondaryText", children: [walletId ? (_jsx(WalletImage, { id: walletId, size: iconSize.md, client: props.client })) : (_jsx(GenericWalletIcon, { size: iconSize.md })), _jsx(Text, { size: "sm", color: "primaryText", children: shortenAddress(props.address) })] }), !props.disableChevron && (_jsx(ChevronDownIcon, { width: iconSize.sm, height: iconSize.sm, style: {
                        color: theme.colors.secondaryText,
                    } })), props.checked && (_jsx(CheckIcon, { width: iconSize.md, height: iconSize.md }))] }) }));
}
